lib/dns.js

```diff
@@ -289,31 +289,38 @@ function setServers(servers) {
   // servers cares won't have any servers available for resolution
   const orig = this._handle.getServers();
   const newSet = [];
-  const IPv6RE = /\[(.*)\]/;
   const addrSplitRE = /(^.+?)(?::(\d+))?$/;
 
   servers.forEach((serv) => {
+    // Easily recognized?
     var ipVersion = isIP(serv);
     if (ipVersion !== 0)
       return newSet.push([ipVersion, serv, IANA_DNS_PORT]);
 
-    const match = serv.match(IPv6RE);
-    // we have an IPv6 in brackets
-    if (match) {
-      ipVersion = isIP(match[1]);
+    // IPv6: /\[(.*)\]/ but O(n) complexity
+    const openBracketIx = serv.indexOf('[');
+    const nextCloseBracketIx = serv.indexOf(']', openBracketIx);
+    if (0 <= openBracketIx && openBracketIx < nextCloseBracketIx) {
+      const withinBrackets = serv.slice(openBracketIx + 1, nextCloseBracketIx);
+      ipVersion = isIP(withinBrackets);
       if (ipVersion !== 0) {
         const port =
           parseInt(serv.replace(addrSplitRE, '$2')) ||
           IANA_DNS_PORT;
-        return newSet.push([ipVersion, match[1], port]);
+        return newSet.push([ipVersion, withinBrackets, port]);
       }
     }
 
-    const [, s, p] = serv.match(addrSplitRE);
-    ipVersion = isIP(s);
+    // addr::port
+    const addrSplitMatch = serv.match(addrSplitRE);
+    if (addrSplitMatch) {
+      const hostIP = addrSplitMatch[1];
+      const port = addrSplitMatch[2] || IANA_DNS_PORT;
```

Comment:
Can you explain the default value here? That is something unexpected for me.

Reply:
`addrSplitMatch[2]` might be undefined (see #20441), in which case I figured the port should be 53 (DNS port).

I think this is consistent with existing cases where a default port is selected, e.g. [here](https://github.com/davisjam/node/blob/404f067c358d712f7090c808b2b0e27351afcd12/lib/dns.js#L298) and [here](https://github.com/davisjam/node/blob/404f067c358d712f7090c808b2b0e27351afcd12/lib/dns.js#L308).

Reply:
@bnoordhuis @apapirovski PTAL

Reply:
I think it should be fine.  `parseInt('')` works out to 0 when coerced to int32 and c-ares uses the default port in that case.  IOW, there should be no observable change in behavior, it'll use port 53 either way.